pkg/policy/mapstate_test.go

```diff
@@ -727,57 +739,338 @@ func (ds *PolicyTestSuite) TestMapState_DenyPreferredInsert(c *check.C) {
-		return TestKey(id, port, proto, trafficdirection.Ingress)
-	}
-	TestEgressKey := func(id int, port uint16, proto uint8) Key {
-		return TestKey(id, port, proto, trafficdirection.Egress)
-	}
-	DNSUDPEgressKey := func(id int) Key {
-		return TestEgressKey(id, 53, 17)
+func testKey(id int, port uint16, proto uint8, direction trafficdirection.TrafficDirection) Key {
+	return Key{
+		Identity:         uint32(id),
+		DestPort:         port,
+		Nexthdr:          proto,
+		TrafficDirection: direction.Uint8(),
 	}
-	DNSTCPEgressKey := func(id int) Key {
-		return TestEgressKey(id, 53, 6)
+}
+
+func testIngressKey(id int, port uint16, proto uint8) Key {
+	return testKey(id, port, proto, trafficdirection.Ingress)
+}
+
+func testEgressKey(id int, port uint16, proto uint8) Key {
+	return testKey(id, port, proto, trafficdirection.Egress)
+}
+
+func DNSUDPEgressKey(id int) Key {
+	return testEgressKey(id, 53, 17)
+}
+
+func DNSTCPEgressKey(id int) Key {
+	return testEgressKey(id, 53, 6)
+}
+
+func HostIngressKey() Key {
+	return testIngressKey(1, 0, 0)
+}
+
+func AnyIngressKey() Key {
+	return testIngressKey(0, 0, 0)
+}
+
+func AnyEgressKey() Key {
+	return testEgressKey(0, 0, 0)
+}
+
+func HttpIngressKey(id int) Key {
+	return testIngressKey(id, 80, 6)
+}
+
+func HttpEgressKey(id int) Key {
+	return testEgressKey(id, 80, 6)
+}
+
+func testEntry(proxyPort uint16, deny bool, owners ...MapStateOwner) MapStateEntry {
+	entry := MapStateEntry{
+		ProxyPort: proxyPort,
+		IsDeny:    deny,
 	}
-	HostIngressKey := func() Key {
-		return TestIngressKey(1, 0, 0)
+	if len(owners) > 0 {
+		entry.owners = make(map[MapStateOwner]struct{}, len(owners))
 	}
-	AnyIngressKey := func() Key {
-		return TestIngressKey(0, 0, 0)
+	for _, owner := range owners {
+		entry.owners[owner] = struct{}{}
 	}
-	//AnyEgressKey := func() Key {
-	//	return TestEgressKey(0, 0, 0)
-	//}
-	HttpIngressKey := func(id int) Key {
-		return TestIngressKey(id, 80, 6)
+	return entry
+}
+
+func testEntryD(proxyPort uint16, deny bool, derivedFrom labels.LabelArrayList, owners ...MapStateOwner) MapStateEntry {
+	entry := testEntry(proxyPort, deny, owners...)
+	entry.DerivedFromRules = derivedFrom
+	return entry
+}
+
+func (ds *PolicyTestSuite) TestMapState_AccumulateMapChangesDeny(c *check.C) {
+	csFoo := newTestCachedSelector("Foo", false)
+	csBar := newTestCachedSelector("Bar", false)
+
+	type args struct {
+		cs       *testCachedSelector
+		adds     []int
+		deletes  []int
+		port     uint16
+		proto    uint8
+		ingress  bool
+		redirect bool
+		deny     bool
 	}
-	HttpEgressKey := func(id int) Key {
-		return TestEgressKey(id, 80, 6)
+	tests := []struct {
+		continued bool // Start from the end state of the previous test
+		name      string
+		setup     MapState
+		args      []args // changes applied, in order
+		state     MapState
+		adds      MapState
+		deletes   MapState
+	}{{
+		name: "test-1a - Adding identity to an existing state",
```

Comment:
nit:
```suggestion
		name: "test-1a - Add L3+L4 deny to existing Allow-all, L4-only allow rule map state entries",
```

Reply:
+1

Reply:
Changed title to `test-1a - Adding L3-deny to an existing allow-all with L4-only allow redirect map state entries` - the L3/L4 deny is then added due to the L4-allow still being in effect for all other L3's.